Add k8s-native IAMStorageAPI

erlendvollset opened the pull request:
## Description
Implements an IAMStorageAPI using annotations on kubernetes config maps.

## Motivation and Context
https://github.com/minio/minio/issues/12338

## How to test this PR?
- Install minikube and set it as current context
- Set following envvars:
```
MINIO_K8S_IAM_STORE_ENABLED=true
MINIO_K8S_KUBE_CONFIG_PATH=~/.kube/config

Optional (set to defaults here):
MINIO_K8S_IAM_STORE_NAMESPACE=default
MINIO_K8S_IAM_STORE_CONFIG_MAP_NAME=minio-k8s-iam-store
```
- start minio

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [X] Unit tests added/updated

erlendvollset committed:
First take on k8s-native config store

erlendvollset committed:
Hardcode config and use k8s store instead of object store by default

erlendvollset committed:
Fix bugs

erlendvollset committed:
Don't harcode config

vadmeste committed:
Update bloom module (#12383)

To fix dependency import issues when importing madmin-go v0.7.1

Commit:
etag: compute ETag as MD5 for compressed single-part objects (#12375)

This commit fixes a bug causing the MinIO server to compute
the ETag of a single-part object as MD5 of the compressed
content - not as MD5 of the actual content.

This usually does not affect clients since the MinIO appended
a `-1` to indicate that the ETag belongs to a multipart object.
However, this behavior was problematic since:
 - A S3 client being very strict should reject such an ETag since
   the client uploaded the object via single-part API but got
   a multipart ETag that is not the content MD5.
 - The MinIO server leaks (via the ETag) that it compressed the
   object.

This commit addresses both cases. Now, the MinIO server returns
an ETag equal to the content MD5 for single-part objects that got
compressed.

Signed-off-by: Andreas Auernhammer <aead@mail.de>

erlendvollset committed:
Removed expired items and fix bug in saveIAMConfig

erlendvollset committed:
fmt

vadmeste committed:
audit/logger: Increase http request timeout (#12385)

A configured audit logger or HTTP logger is validated during MinIO
server startup. Relax the timeout to 10 seconds in that case, otherwise,
both loggers won't be used.

1 second could be too low for a busy HTTP endpoint.

harshavardhana committed:
fix: always prefer DataBlocks present in FileInfo (#12386)

harshavardhana committed:
feat: introduce listUsers, listPolicies for any bucket (#12372)

Bonus change LDAP settings such as user, group mappings
are now listed as part of `mc admin user list` and
`mc admin group list`

Additionally this PR also deprecates the `/v2` API
that is no longer in use.

andersfylling commented on cmd/config/k8s/k8s.go:
```diff
@@ -0,0 +1,132 @@
+		}
+	}
+	clientset, err := kubernetes.NewForConfig(k8sconfig)
+	if err != nil {
```
No need to validate err. You can just return the method products `return kubernetes.NewForConfig(k8sconfig)`

andersfylling commented on cmd/config/k8s/k8s.go:
```diff
@@ -0,0 +1,132 @@
+	var k8sconfig *rest.Config
+	var err error
+	if cfg.KubeConfigPath != "" {
+		k8sconfig, err = clientcmd.BuildConfigFromFlags("", cfg.KubeConfigPath)
```
I notice that this seems to be frequently used:
```go
if k8sconfig, err = clientcmd.BuildConfigFromFlags("", cfg.KubeConfigPath); err != nil {
    return nil, err
}
```

andersfylling commented on cmd/config/k8s/k8s.go:
```diff
@@ -0,0 +1,132 @@
+			return nil, err
+		}
+	} else {
+		k8sconfig, err = rest.InClusterConfig()
```
I notice that this seems to be frequently used:
```go
if k8sconfig, err = rest.InClusterConfig(); err != nil {
    return nil, err
}
```

andersfylling commented on cmd/config/k8s/k8s.go:
```diff
@@ -0,0 +1,132 @@
+		return nil, nil
+	}
+	var k8sconfig *rest.Config
+	var err error
```
you can use named return values to remove this line https://tour.golang.org/basics/7

`func New(cfg Config) (*kubernetes.Clientset, error) {` => `func New(cfg Config) (clientset *kubernetes.Clientset, err error) {`

But I am uncertain what is preferred in this repo.

erlendvollset replied:
Ooh. Nice feature. Didn't know about that one!

klauspost committed:
Add more erasure codes on degraded systems. (#11852)

In cases where a cluster is degraded, we do not uphold our consistency 
guarantee and we will write fewer erasure codes and rely on healing 
to recreate the missing shards.

In some cases replacing known bad disks in practice take days.
We want to change the behavior of a known degraded system to keep
the erasure code promise of the storage class for each object.

This will create the objects with the same confidence as a fully 
functional cluster. The tradeoff will be that objects created 
during a partial outage will take up slightly more space.

This means that when the storage class is EC:4, there should 
always be written 4 parity shards, even if some disks are unavailable.

When an object is created on a set, the disks are immediately 
checked. If any disks are unavailable additional parity shards 
will be made for each offline disk, up to 50% of the number of disks.

We add an internal metadata field with the actual and intended 
erasure code level, this can optionally be picked up later by 
the scanner if we decide that data like this should be re-sharded.

andersfylling commented on cmd/config-current.go:
```diff
@@ -395,6 +409,18 @@ func lookupConfigs(s config.Config, setDriveCounts []int) {
 		}
 	}
 
+	k8sCfg, err := k8s.LookupConfig(s[config.K8sSubSys][config.Default])
```
I wish there was a re-usable way to setup the instance here, that could be shared with etcd and xldap, but but it might make everything harder to read.

I would move `globalK8sIamStoreConfig = k8sCfg` above the `New` call to make the behaviour intent clearer.

erlendvollset replied:
> I wish there was a re-usable way to setup the instance here, that could be shared with etcd and xldap, but but it might make everything harder to read.

Agree. Probably best to leave such a refactoring to the minio guys though, they probably have some opinions on how it should be done.

> I would move globalK8sIamStoreConfig = k8sCfg above the New call to make the behaviour intent clearer.

👍

andersfylling commented on cmd/iam-k8s-store.go:
```diff
@@ -0,0 +1,554 @@
+		return err
+	}
+	attempts := 0
+	for attempts < iamK8s.maxUpdateAttempts {
```
This for loop is large. Could the content be broken down into functions?

andersfylling commented on cmd/iam-k8s-store.go:
```diff
@@ -0,0 +1,554 @@
+			},
+		}
+		_, err = iamK8s.configMapsClient.Update(ctx, configMapUpdated, metav1.UpdateOptions{})
+		if err != nil {
```
I wonder if the k8s error module uses error.As which returns false when err is nil. You also have a very similar err check in a later function, I would suggest merging them / re-use the logic if it's possible. But yes, I don't think you need to wrap this in a "err != nil", but I can't guarantee it.

andersfylling commented on cmd/iam-k8s-store.go:
```diff
@@ -0,0 +1,554 @@
+	var p iampolicy.Policy
+	err := iamK8s.loadIAMConfig(ctx, &p, getPolicyDocPath(policy))
+	if err != nil {
+		if err == errConfigNotFound {
```
Would perhaps wrapping the error be better? Also I would make error instances (errNoSuchPolicy) public such that people can match against them, and programmatically deal with different scenarios, in general.

But that only applies if users will ever see error results from this method, or any private methods in this package/struct.

andersfylling commented on cmd/iam-k8s-store.go:
```diff
@@ -0,0 +1,554 @@
+	err := iamK8s.loadIAMConfig(ctx, &u, getUserIdentityPath(user, userType))
+	if err != nil {
+		if err == errConfigNotFound {
+			return errNoSuchUser
```
same comment as above

andersfylling commented on cmd/iam-k8s-store.go:
```diff
@@ -0,0 +1,554 @@
+	for _, item := range items {
+		if hasTtlPrefix(item.objPath) {
+			expTime, _ := strconv.ParseInt(item.data, 10, 64)
+			if expTime > time.Now().Unix() {
```
Potential race where some entries might be treated as expired when they have the same timestamp as others. I would move `time.Now().Unix()` to a variable before the for loop starts.

Is it likely to happen? no. I experienced something similar earlier, but I'm uncertain if this matters for your case.

andersfylling commented on cmd/iam-k8s-store.go:
```diff
@@ -0,0 +1,554 @@
+			logger.Info("Succesfully removed " + removedItems + " expired items from iam config store")
+			return
+		}
+		if errorsv1.IsConflict(err) {
```
this looks very familiar

andersfylling reviewed (approved):
LGTM! Just some nitpicking and ideas, but nothing critical.

harshavardhana committed:
fix: when parityDrives hits > len(storageDisks)/2, keep maxParity (#12387)

Additionally move out `x-minio-internal-erasure-upgraded` from HTTP headers
list, as its an internal header, rename elsewhere accordingly.

minio-trusted committed:
Update yaml files to latest version RELEASE.2021-05-27T22-06-31Z

erlendvollset committed:
Address comments

erlendvollset committed:
Fix potential race

erlendvollset committed:
fmt

erlendvollset committed:
Merge branch 'master' into k8s-native-iam-store

erlendvollset committed:
Introduce interface k8sConfigMapClient interface so we can mock out k8s calls in tests

erlendvollset committed:
Merge branch 'k8s-native-iam-store' of github.com:cognitedata/minio into k8s-native-iam-store

erlendvollset committed:
lint

erlendvollset committed:
Add some more unit tests

erlendvollset committed:
lint

erlendvollset committed:
Fix error returned from loadIAMConfig and use Sprintf
